api/src/main/java/org/pmiops/workbench/reporting/ReportingSnapshotServiceImpl.java

```diff
@@ -1,45 +1,68 @@
 package org.pmiops.workbench.reporting;
 
+import com.google.common.base.Stopwatch;
 import java.time.Clock;
 import java.util.List;
 import java.util.Random;
+import java.util.logging.Logger;
+import javax.inject.Provider;
 import org.pmiops.workbench.db.dao.UserService;
 import org.pmiops.workbench.db.model.DbUser;
 import org.pmiops.workbench.db.model.DbWorkspace;
 import org.pmiops.workbench.model.ReportingResearcher;
 import org.pmiops.workbench.model.ReportingSnapshot;
 import org.pmiops.workbench.model.ReportingWorkspace;
 import org.pmiops.workbench.workspaces.WorkspaceService;
+import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.stereotype.Service;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.support.TransactionTemplate;
 
 @Service
 public class ReportingSnapshotServiceImpl implements ReportingSnapshotService {
+  private static final Logger log = Logger.getLogger(ReportingSnapshotServiceImpl.class.getName());
 
   private final Clock clock;
   private final ReportingMapper reportingMapper;
-  private Random random;
+  private final Random random;
+  private final PlatformTransactionManager platformTransactionManager;
+  private final Provider<Stopwatch> stopwatchProvider;
   private final UserService userService;
   private final WorkspaceService workspaceService;
 
   public ReportingSnapshotServiceImpl(
       Clock clock,
       ReportingMapper reportingMapper,
       Random random,
+      @Qualifier("transactionManager") PlatformTransactionManager platformTransactionManager,
+      Provider<Stopwatch> stopwatchProvider,
       UserService userService,
       WorkspaceService workspaceService) {
     this.clock = clock;
     this.reportingMapper = reportingMapper;
     this.random = random;
+    this.platformTransactionManager = platformTransactionManager;
+    this.stopwatchProvider = stopwatchProvider;
     this.userService = userService;
     this.workspaceService = workspaceService;
   }
 
   @Override
   public ReportingSnapshot takeSnapshot() {
-    return new ReportingSnapshot()
-        .captureTimestamp(clock.millis())
-        .researchers(getResearchers())
-        .workspaces(getWorkspaces());
+    final Stopwatch stopwatch = stopwatchProvider.get().start();
+    final TransactionTemplate template = new TransactionTemplate(platformTransactionManager);
```

Comment:
Flagging this: needs review by someone more Spring savvy

Reply:
I expect to revisit it when I do the [perf testing ticket](https://precisionmedicineinitiative.atlassian.net/browse/RW-5196). I think it's going to be important to schedule a couple of these soon after the initial delivery if we can.

Is there a  way to flag it as read-only if I don't use the template stuff?

Reply:
@Transactional(readOnly = true)